x/orderbook/model.go

```diff
@@ -0,0 +1,270 @@
+	if o.OriginalOffer == nil {
+		return errors.Wrap(errors.ErrEmpty, "original offer")
+	}
+	if err := o.OriginalOffer.Validate(); err != nil {
+		return errors.Wrap(err, "original offer")
+	}
+	if o.RemainingOffer == nil {
+		return errors.Wrap(errors.ErrEmpty, "remaining offer")
+	}
+	if err := o.RemainingOffer.Validate(); err != nil {
+		return errors.Wrap(err, "remaining offer")
+	}
+	if err := o.Price.Validate(); err != nil {
+		return errors.Wrap(err, "price")
+	}
+	if !o.Price.IsPositive() {
+		return errors.Wrap(errors.ErrState, "price must be positive")
+	}
+	// TODO: valid trade ids (also rethink how we handle this? just use index and not in model?)
+
+	if err := o.UpdatedAt.Validate(); err != nil {
+		return errors.Wrap(err, "updated at")
+	}
+	if o.UpdatedAt == 0 {
+		return errors.Wrap(errors.ErrEmpty, "missing updated at")
+	}
+	if err := o.CreatedAt.Validate(); err != nil {
+		return errors.Wrap(err, "created at")
+	}
+	if o.CreatedAt == 0 {
+		return errors.Wrap(errors.ErrEmpty, "missing created at")
+	}
+	return nil
+}
+
+var _ morm.Model = (*Trade)(nil)
+
+// SetID is a minimal implementation, useful when the ID is a separate protobuf field
+func (t *Trade) SetID(id []byte) error {
+	t.ID = id
+	return nil
+}
+
+// Copy produces a new copy to fulfill the Model interface
+func (t *Trade) Copy() orm.CloneableData {
+	return &Trade{
+		Metadata:    t.Metadata.Copy(),
+		ID:          copyBytes(t.ID),
+		OrderBookID: copyBytes(t.OrderBookID),
+		OrderID:     copyBytes(t.OrderID),
+		Taker:       copyBytes(t.Taker),
+		Maker:       copyBytes(t.Maker),
+		MakerPaid:   t.MakerPaid.Clone(),
+		TakerPaid:   t.TakerPaid.Clone(),
+		ExecutedAt:  t.ExecutedAt,
+	}
+}
+
+// Validate is always succesful
+func (o *Trade) Validate() error {
+	if err := o.Metadata.Validate(); err != nil {
+		return err
+	}
+	if err := isGenID(o.ID, true); err != nil {
+		return err
+	}
+	if err := isGenID(o.OrderBookID, false); err != nil {
+		return errors.Wrap(err, "order book id")
+	}
+	if err := isGenID(o.OrderID, false); err != nil {
+		return errors.Wrap(err, "order  id")
+	}
+	if err := o.Taker.Validate(); err != nil {
+		return errors.Wrap(err, "taker")
+	}
+	if err := o.Maker.Validate(); err != nil {
+		return errors.Wrap(err, "maker")
+	}
+	if o.MakerPaid == nil {
+		return errors.Wrap(errors.ErrEmpty, "maker paid")
+	}
+	if err := o.MakerPaid.Validate(); err != nil {
+		return errors.Wrap(err, "maker paid")
+	}
+	if o.TakerPaid == nil {
+		return errors.Wrap(errors.ErrEmpty, "maker paid")
+	}
+	if err := o.TakerPaid.Validate(); err != nil {
+		return errors.Wrap(err, "taker paid")
+	}
+	if err := o.ExecutedAt.Validate(); err != nil {
+		return errors.Wrap(err, "executed at")
+	}
+	if o.ExecutedAt == 0 {
+		return errors.Wrap(errors.ErrEmpty, "missing executed at")
+	}
+	return nil
+}
+
+// isGenID ensures that the ID is 8 byte input.
+// if allowEmpty is set, we also allow empty
+func isGenID(id []byte, allowEmpty bool) error {
```

Comment:
Why would we want to allow empty GenId?

Reply:
`isGenID` is just that it is the valid format for the IDGenBucket.
When we first save an object, we have an empty id, which is auto-generated by the ModelBucket (auto-incrementing sequence). However, we validate before adding an ID, so we should allow this to be missing.

Maybe we should assign the ID *before* Validating and then we can require it?

Reply:
> Maybe we should assign the ID before Validating and then we can require it?

Please feel free to bring this up. when we write Buckets and Handlers